### ISSUE TITLE
Setuptools 25 doesn't handle "pip install -e" over the top of existing "pip install"

### ISSUE THREAD

**Issue by therefromhere** [pypa/setuptools#729]
When editing packages I've previous done "pip install -e" without first doing "pip uninstall" - this workflow is broken with setuptools 25 - pip freeze reports the wrong (non -e version), and pip uninstall fails with this error: 

```
$ pip uninstall simplejson
Exception:
Traceback (most recent call last):
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/commands/uninstall.py", line 76, in run
    requirement_set.uninstall(auto_confirm=options.yes)
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/req/req_set.py", line 336, in uninstall
    req.uninstall(auto_confirm=auto_confirm)
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/req/req_install.py", line 687, in uninstall
    '(at %s)' % (link_pointer, self.name, dist.location)
AssertionError: Egg-link /tmp/simplejson does not match installed location of simplejson (at /home/johnc/.virtualenvs/bug/lib/python2.7/site-packages)
```

(ignore that I'm using simplejson as an example, I'm getting the same thing with my own packages).

Steps to reproduce:

```
$ mkvirtualenv bug
$ pip install setuptools~=25.1.6
$ pip install simplejson
$ pip freeze
simplejson==3.8.2
$ pip uninstall simplejson
# (uninstall works as expected)
$ pip install simplejson
$ pip freeze
simplejson==3.8.2
# ok, install/uninstall works as expected
$ git clone https://github.com/simplejson/simplejson.git
$ pip install -e simplejson/
Obtaining file:///tmp/simplejson
Installing collected packages: simplejson
  Running setup.py develop for simplejson
Successfully installed simplejson-3.8.2
$ pip freeze
simplejson==3.8.2
# ??????? expected an editiable mode git hash
$ pip uninstall simplejson
Exception:
Traceback (most recent call last):
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/commands/uninstall.py", line 76, in run
    requirement_set.uninstall(auto_confirm=options.yes)
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/req/req_set.py", line 336, in uninstall
    req.uninstall(auto_confirm=auto_confirm)
  File "/home/johnc/.virtualenvs/bug/local/lib/python2.7/site-packages/pip/req/req_install.py", line 687, in uninstall
    '(at %s)' % (link_pointer, self.name, dist.location)
AssertionError: Egg-link /tmp/simplejson does not match installed location of simplejson (at /home/johnc/.virtualenvs/bug/lib/python2.7/site-packages)
```

By contrast, this works correctly with an earlier version of setuptools.  Expected behaviour:

```
$ mkvirtualenv notbug
$ pip install setuptools~=24.3.1
$ pip install simplejson
$ pip freeze
simplejson==3.8.2
$ pip uninstall simplejson
# (uninstall works as expected)
$ pip install simplejson
$ pip freeze
simplejson==3.8.2
# ok, install/uninstall works as expected
$ git clone https://github.com/simplejson/simplejson.git
$ pip install -e simplejson/
Obtaining file:///tmp/simplejson
Installing collected packages: simplejson
  Running setup.py develop for simplejson
Successfully installed simplejson-3.8.2
$ pip freeze
-e git+https://github.com/simplejson/simplejson.git@4b1dab370f8a32aaa3f9140c38645093d3287b82#egg=simplejson
# editable mode as expected
$ pip uninstall simplejson
Uninstalling simplejson-3.8.2:
  /home/johnc/.virtualenvs/notbug/lib/python2.7/site-packages/simplejson.egg-link
Proceed (y/n)? y
  Successfully uninstalled simplejson-3.8.2
$ pip freeze
simplejson==3.8.2
# and we're back to the normal package, as expected
```

**Comment by jaraco** [comment]
The change in Setuptools 25 is the removal of sys path reordering. This issue might be a dupe of #447. @minrk Can you investigate?

**Comment by minrk** [comment]
I think this is a dupe of #447, and a fix is already open on pip: https://github.com/pypa/pip/pull/3898. The pip fix wouldn't address people who type `setup.py develop` directly, though, which will have this problem unless either

1. develop does its own uninstall, like pip, or
2. develop is somehow special-cased to continue taking higher priority, without affecting 'regular' installs

There does seem to be a bug in pip, though, regarding the uninstall not correctly handling the fact that the egg-link will have lower import priority than the existing install. I'm guessing there's an assumption about the old setuptools behavior somewhere in pip's code for selecting which package is active.

**Comment by minrk** [comment]
The uninstall failure should be fixed by https://github.com/pypa/pip/pull/3904

**Comment by faassen** [comment]
See here for my experience with this bug: https://github.com/pypa/pip/pull/3904#issuecomment-242339808

**Comment by therefromhere** [comment]
Upgrading to pip 9 seems to have resolved this (due to pypa/pip#3898)
